### PR TITLE
[POC][sqlite] CRSQLite integration

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Completes ENG-8809
Investigate possibility of using [crsqlite](https://vlcn.io/) in `expo-sqlite` which allows database syncing between clients.

# How
- Added the crsqlite extension as a framework and provided an event that contains the changeset to send to the server. 
- Currently only works on iOS.
- Needs input on an API
- Discussion about how to handle the sync server
- Need to look into demo using TinyBase. This will require writing a custom `persister`

# Test Plan
Created a demo project and a small sync server to test the sync on multiple clients

https://github.com/expo/expo/assets/30924086/fe98eea0-f0a2-4541-88a5-080098983d3a

### CODE REVIEW

**Comment by tantaman** [comment]
Awesome work!

Some feedback, thoughts and related items. Not everything that follows is a priority item but I wanted to provide as much relevant context as possible to the Expo team --

# Expo SQLite Bindings

While working with @alanjhughes, we discovered that the expo bindings need quite a bit of work 😬. The immediate issues that jump out:
1. Expo has no support for prepared statements
2. Expo has no support for passing binary (e.g., Uint8Array) data back and forth

(1) is a big perf hit for apps. When `executing` an unprepared query, 50-75% of your query time will be spent preparing the statement. Allowing devs to prepare and re-use their statement gives some easy and immediate perf wins.

Statement preparation is also a big hit on battery life as it is a very CPU intensive process.

(2) makes us jump through a bunch of hoops converting binary data to hex strings and back. This is error prone and an annoying developer experience.

# Relevant CR-SQLite Roadmap Items

We have a number of issues coming to a close in the next few weeks that are relevant to production readiness.

1. Reducing metadata size. The CRDT metadata that I track can be reduced by around 5x so it is much less of the total database weight (https://github.com/vlcn-io/cr-sqlite/issues/293)
2. Re-insertion. cr-sqlite currently does not let users re-insert rows that were previously deleted. This makes creating apps with undo functionality a pain and also increases CRDT metadata size. This'll be fixed soon thanks to some support from fly.io (https://github.com/vlcn-io/cr-sqlite/issues/71)

I'm sure it'll be a while before Expo actually ships the integration with cr-sqlite to users but, in any case, I'd wait to ship until these two items are addressed given they would be breaking changes to the API.

Two other items that are a bit further out --
1. We have an algorithm sketched out (but not yet implemented) to support collaborative text documents in cr-sqlite -- https://github.com/vlcn-io/cr-sqlite/issues/65#issuecomment-1591947922. The algorithm is [Fugue](https://arxiv.org/abs/2305.00583), adapted to work with relations. The idea would be that app developers can maintain collaborative text docs directly in SQLite without needing a separate library like yjs or automerge. Credit goes to @mweidner037 for this.
2. We have an improved syntax for "conflict free table creation" planned out -- https://github.com/vlcn-io/cr-sqlite/issues/181. Ty to @asg017 for the clever trick that makes this possible.

These last two items should not introduce any breaking changes.

# Sync Server

I don't have a service that people can just point to and use to facilitate sync. I do, however, have a [sync server implementation](https://github.com/vlcn-io/cr-sqlite/tree/main/js/packages/direct-connect-nodejs) that users can self host. This is what the [vite starter uses](https://vite-starter.fly.dev/), for example. [strut.io](https://strut.io) also uses that implementation. Strut is interesting as it saves the SQLite DBs on [litefs](https://fly.io/blog/introducing-litefs/) to support realtime geo-distributed replicas which are strongly consistent.

Other options --

- sqld & turso. I'm working on integrating cr-sqlite into https://turso.tech. The current blocker is a lack of multi-tenancy but that is being fixed. Maybe there could be some partnership here with Expo to allow Expo users to easily spin up a Turso DB for sync? cc @glommer
- @conradev had some interesting ideas on syncing through bluetooth, wifi-direct or other "local" networking options. Not sure if he's implemented anything there yet or not though.

# TinyBase

I currently consider the tinybase integration to be in a "proof of concept" state (although I'm curious for @jamesgpearce's thoughts). Ideally there'd be some work done here on making the communication between TinyBase and SQLite incremental before advertising it as ready for production use cases.

## Developer API - Typed-SQL & TreeSQL

This is related to "how do we best expose the database for developers to interact with it?" The following is a bit far afield from the integration work itself but I mention it in the hope of finding like minded people to push this style of development forward and/or sway the direction of Expo wrt surfacing SQLite.

I'm in favor of removing layers between the App and the DB itself. IMO, people generally reach for an ORM (or some other abstraction over their DB) to help with:

1. Migrations
2. End to end types
3. Pulling data out in a tree structure rather than flat structure

These can all be solved without an ORM.

I've been teaching TypeScript to understand SQL here --

https://github.com/vlcn-io/typed-sql

And teaching SQL to understand hierarchical data here (this is a _very rough_ 100 SLOC prototype) --

https://github.com/tantaman/TreeSQL

And [implementing](https://github.com/vlcn-io/cr-sqlite/blob/main/core/rs/core/src/automigrate.rs) [this basic idea](https://david.rothlis.net/declarative-schema-migration-for-sqlite/) for automatic SQLite migrations

This is quite a large scope of work for one person who is also spending most of his time supporting `cr-sqlite` which is why I bring it up in the hope of finding like-minded collaborators.

# Future future

There are some other cool ideas I've been exploring with the Riffle folks with respect to getting 60FPS while keeping every single bit of application state in the SQLite database.

Much of this has been very pragmatic work to squeeze out as much perf from SQLite and React as possible although we're starting to narrow in on generalized concepts and a form of reactive SQL.

File: apps/test-suite/tests/SQLite.js [hunk 1]
```diff
@@ -508,6 +508,19 @@ export function test(t) {
         });
       });
 
+      t.it('should load crsqlite extension correctly', async () => {
+        const db = SQLite.openDatabase('test.db');
+        await db.execAsync([{ sql: 'DROP TABLE IF EXISTS foo;', args: [] }], false);
```
Comment (tantaman) [inline apps/test-suite/tests/SQLite.js:513]: nit: technically you should

```
SELECT crsql_begin_alter('foo');
DROP TABLE foo;
SELECT crsql_commit_alter('foo');
```

when modifying crrs. The `select` to start modification is an eye-sore but something that I'll address in https://github.com/vlcn-io/cr-sqlite/issues/181

File: packages/expo-sqlite/ios/SQLiteModule.swift [hunk 2]
```diff
@@ -51,8 +64,17 @@ public final class SQLiteModule: Module {
...
+      hasListeners = true
+    }
+
+    OnStopObserving {
+      hasListeners = false
+    }
+
     OnDestroy {
       cachedDatabases.values.forEach {
+        executeSql(sql: "SELECT crsql_finalize()", with: [], for: $0, readOnly: false)
```
Comment (tantaman) [inline packages/expo-sqlite/ios/SQLiteModule.swift:77]: 👍

Might be worth pushing this sort of detail down into the SQLite bindings themselves so users can never forget to `finalize` the extension.

fwiw -- the libsql fork of sqlite has fixed the need to do this: https://github.com/libsql/libsql/issues/62

File: packages/expo-sqlite/src/SQLite.ts [hunk 3]
```diff
@@ -35,12 +36,12 @@ export class SQLiteDatabase {
   /**
    * Executes the SQL statement and returns a callback resolving with the result.
    */
-  exec(queries: Query[], readOnly: boolean, callback: SQLiteCallback): void {
+  exec(queries: Query[], readOnly: boolean, requiresSync: boolean, callback: SQLiteCallback): void {
```
Comment (tantaman) [inline packages/expo-sqlite/src/SQLite.ts:39]: imo you guys would want to figure out some way to remove any flags related to `requiresSync`

One option is to use two separate connections for writes. One for writes from the sync layer, one for writes from the app itself. This way you can differentiate between remote and local events and it can be hidden away from developers that don't need to know those details but still available for those that do.

**Comment by jozan** [comment]
this is great!

regarding prepared statements. seems that all sql queries are using prepared statements on both [ios](https://github.com/expo/expo/blob/47a99a8015e9eef20e93ae2530c5b8a37d7c205c/packages/expo-sqlite/ios/SQLiteModule.swift#L102C10-L102C10) and [android](https://github.com/expo/expo/blob/47a99a8015e9eef20e93ae2530c5b8a37d7c205c/packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModule.kt#L74C1-L74C1). am i missing something?

there has been pending feature request for [prepared statements since 2017](https://expo.canny.io/feature-requests/p/sqlite-support-for-prepared-statements). this is probably obsolete.

**Comment by tantaman** [comment]
> regarding prepared statements. seems that all sql queries are using prepared statements on both [ios](https://github.com/expo/expo/blob/47a99a8015e9eef20e93ae2530c5b8a37d7c205c/packages/expo-sqlite/ios/SQLiteModule.swift#L102C10-L102C10) and [android](https://github.com/expo/expo/blob/47a99a8015e9eef20e93ae2530c5b8a37d7c205c/packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModule.kt#L74C1-L74C1). am i missing something?


They are prepared inside of `exec` as you linked.

The problem is that there is no way (at least glancing at API docs -- https://docs.expo.dev/versions/latest/sdk/sqlite/) for the user of your bindings to prepare the statement themselves.

What this means is that each call to `exec` re-prepares the query rather than re-using the prepared query.

`sqlite3_prepare_v2` is rather expensive as it does much more than just parsing the SQL string. It also creates many possible execution plans and tests them to figure out which plan is optimal.

So instead of the API you have now:

```ts
const results1 = sqlite.exec(`SELECT * FROM foo WHERE x = ?`, [arg]);
const results2 = sqlite.exec(`SELECT * FROM foo WHERE x = ?`, [arg]);
const results3 = sqlite.exec(`SELECT * FROM foo WHERE x = ?`, [arg]);
```

You'd have:

```ts
const prepared = sqlite.prepare(`SELECT * FROM foo WHERE x = ?`);
const results1 = prepared.fetchAll([arg]);
const results2 = prepared.fetchAll([arg]);
const results3 = prepared.fetchAll([arg]);
```

In the latter case, each call to `fetchAll` gets the benefit of not having to do the preparation step & saves tons of time and CPU.

If you want to go deeper on all the work that happens in statement prep -- https://fly.io/blog/sqlite-virtual-machine/

**Comment by jozan** [comment]
oh, thanks for explaining in detail!

> The problem is that there is no way (at least glancing at API docs -- https://docs.expo.dev/versions/latest/sdk/sqlite/) for the user of your bindings to prepare the statement themselves.
> 
> What this means is that each call to `exec` re-prepares the query rather than re-using the prepared query.
> 

i checked the code and there is no way to create prepared statements separately at all. i actually stumbled on this today when i was  making poc to get drizzle orm running on `expo-sqlite`. their every sqlite binding is built on top of prepared statements.

**Comment by jozan** [comment]
some good news. expo's very own [james and brent responded](https://twitter.com/notbrent/status/1684333448549527552) to my tweets about current state of expo-sqlite and told that they are actually working on some needed improvements!

hopefully it'll help out your experiments with crsqlite to move forward.

**Comment by jamesgpearce** [comment]
To do a 'proper' TinyBase integration, we will ideally need a way to get to sqlite3_update_hook. Without it, TinyBase won't know when data has changed and we're stuck with manual loading and/or polling.

**Comment by tantaman** [comment]
> To do a 'proper' TinyBase integration, we will ideally need a way to get to sqlite3_update_hook. Without it, TinyBase won't know when data has changed and we're stuck with manual loading and/or polling.

Right, I guess the expo bindings are missing `sqlite3_update_hook` as well.

Update notifications could be implemented in the expo bindings themselves given all writes to the DB will go through those bindings.

Some things to note about the [sqlite3_update_hook](https://www.sqlite.org/c3ref/update_hook.html) --

1. It does not call you back if changes are made by [other connections to the DB](https://sqlite.org/forum/forumpost/83cba5700f9d4db6)
2. It calls you back before transaction commit so the change that the update hook notifies you about could get rolled back. I didn't find docs stating this but experimentation shows it to be true.
3. Only _one_ update hook callback can be registered per connection so if the app developer adds their own callback, yours will get removed.
   > Any callback set by a previous call to this function for the same database connection [is overridden](https://www.sqlite.org/c3ref/update_hook.html).
5. sqlite3_update_hook does not notify when changes are made to `WITHOUT ROWID` tables
   > sqlite3_update_hook  ... to be invoked whenever a row is updated, inserted or deleted in a **[rowid table](https://www.sqlite.org/rowidtable.html)**

Given all those issues ([summary](https://sqlite.org/forum/forumpost/cb4f351d86670152)), I'd suggest implementing notifications within the expo bindings and without using the update hook. The bindings would track writes and --

1. Notify all connections of writes made by other connections
2. Withhold notifications until transaction commit
3. Allow many callbacks to be registered for notifications
5. Work regardless of the underlying table type(s) (e.g., rowid or without rowid)

This is something I could help with and/or port over from some of my existing work.

**Comment by tantaman** [comment]
I'm also chatting with the author of https://github.com/groue/GRDB.swift to get some alternate ideas on observing database values.

**Comment by alanjhughes** [comment]
Closed in favour of #23791 now published in an alpha release

**Comment by tantaman** [comment]
I documented some of my discussions with @groue here: https://github.com/vlcn-io/cr-sqlite/discussions/309

tldr -- we can use the update_hook if we place some constraints on users (only allow a single write connection, do not allow WITHOUT ROWID tables) and can get pretty fine grained observability.
